### PR TITLE
Scheduled price rule time zone correction

### LINKED ISSUE #29549: [Issue] Scheduled price rule time zone correction
This issue is automatically created based on existing pull request: magento/magento2#28973: Scheduled price rule time zone correction

### Description (*)

Apply the same logic for time zone correction on `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` as in `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute`.

The problem here is that Magento 2.3 uses the method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule`. In Magento 2.4 this method is no longer used, but it is also not deprecated. So we can expect customers to use this method through the `@api` method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById`. So even though the code in this PR is not used anywhere in vanilla M2.4, it is an `@api` method that could result in catalog rules starting and ending at wrong time which is equal to the configured timezone offset.

Applying this PR makes sure that time zones are applied equally in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById` as they are in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexByIds`. Currently they differ.

### Precondition
Magento 2.4-develop

### Steps to reproduce (*)
1. Set time zone to Europe/Amsterdam
1. Set indexers to scheduled
1. Deleta all present Catalog Price Rules
1. Create a Catalog Price Rule for 50% off on all products, but disabled
1. Schedule the price rule to be enabled at a minute or so in the future using catalog staging
1. Wait until catalog stating rules are active and indexer is up to date
1. Run this query: `SELECT DISTINCT from*time, to_time FROM catalogrule*product;`
1. It should output only one row, after all, the rule is applied on all product at the same time.
1. Install [app.zip](https://github.com/magento/magento2/files/4868883/app.zip)
1. Run `bin/magento reindex:product 10`
1. Run this query: `SELECT DISTINCT from*time, to_time FROM catalogrule*product;`

### Actual Result
Now you see two rows: one from before and one with a time offset 2 hours into the future. The Europe/Amsterdam timezone is not converted to UTC here. This is the issue.

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
--]
`\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` and `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute` do contain duplicate code. Maybe, after approving this PR, it would be worth while to deprecate or merge some of the methods.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x> All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by AntonEvers**
### Description (*)

Apply the same logic for time zone correction on `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` as in `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute`.

The problem here is that Magento 2.3 uses the method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule`. In Magento 2.4 this method is no longer used, but it is also not deprecated. So we can expect customers to use this method through the `@api` method `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById`. So even though the code in this PR is not used anywhere in vanilla M2.4, it is an `@api` method that could result in catalog rules starting and ending at wrong time which is equal to the configured timezone offset.

Applying this PR makes sure that time zones are applied equally in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexById` as they are in `\Magento\CatalogRule\Model\Indexer\IndexBuilder::reindexByIds`. Currently they differ.

### Manual testing scenarios (*)

1. Set time zone to Europe/Amsterdam
2. Set indexers to scheduled
3. Deleta all present Catalog Price Rules
4. Create a Catalog Price Rule for 50% off on all products, but disabled
5. Schedule the price rule to be enabled at a minute or so in the future using catalog staging
6. Wait until catalog stating rules are active and indexer is up to date
7. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`
8. It should output only one row, after all, the rule is applied on all product at the same time.
9. Install [app.zip](https://github.com/magento/magento2/files/4868883/app.zip)
10. Run `bin/magento reindex:product 10`
11. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`
12. Now you see two rows: one from before and one with a time offset 2 hours into the future. The Europe/Amsterdam timezone is not converted to UTC here. This is the issue.

### Questions or comments

`\Magento\CatalogRule\Model\Indexer\IndexBuilder::assignProductToRule` and `\Magento\CatalogRule\Model\Indexer\ReindexRuleProduct::execute` do contain duplicate code. Maybe, after approving this PR, it would be worth while to deprecate or merge some of the methods.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29549: Scheduled price rule time zone correction

### CODE REVIEW

**Comment by AntonEvers** [comment]
@magento run Unit Tests
@magento run Static Tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento create issue

File: app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php [hunk 1]
```diff
@@ -404,16 +418,23 @@ private function assignProductToRule(Rule $rule, int $productEntityId, array $we
...
-        $toTime = strtotime($rule->getToDate());
-        $toTime = $toTime ? $toTime + self::SECONDS_IN_DAY - 1 : 0;
         $sortOrder = (int)$rule->getSortOrder();
         $actionOperator = $rule->getSimpleAction();
         $actionAmount = $rule->getDiscountAmount();
         $actionStop = $rule->getStopRulesProcessing();
 
         $rows = [];
         foreach ($websiteIds as $websiteId) {
+            $scopeTz = new \DateTimeZone(
```
Comment (sidolov) [inline app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php:435]: Please, use framework TimeZone and DateTime object instead of generic PHP.

**Review by sidolov — CHANGES_REQUESTED** [review]
Please, cover your changes with any kind of automated tests.

File: app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php [hunk 2]
```diff
@@ -237,25 +230,18 @@ public function __construct(
...
-            $this->cleanProductIndex([$id]);
-
-            $products = $this->productLoader->getProducts([$id]);
-            $activeRules = $this->getActiveRules();
-            foreach ($products as $product) {
-                $this->applyRules($activeRules, $product);
-            }
-
-            $this->reindexRuleGroupWebsite->execute();
+            $this->doReindexByIds([$id]);
```
Comment (AntonEvers) [inline app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php]: This is the main change in this file, the rest is removal of unused private methods (containing also duplicate code) after this change and adding of imports.

File: app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php (hunk 1 above)
Comment (AntonEvers) [reply app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php:435]: I have removed the entire method, as it appears to be duplicate code. Instead I have done this: https://github.com/magento/magento2/pull/28973/files#r474752002

**Comment by engcom-Charlie** [comment]
Hi, @AntonEvers could you please merge conflict and cover changes with the automated tests?
Thank you.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
> Hi, @AntonEvers could you please merge conflict and cover changes with the automated tests?
> Thank you.

Hi @engcom-Foxtrot, I think we are good to go right? I cannot add tests, as I only removed code. So the tests that are already in place should cover the scenarios that need testing.

There are no conflicts anymore. If anything more needs to be done on this PR, please let me know.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
Hello, @AntonEvers. Thank you for collaboration. 

> So the tests that are already in place should cover the scenarios that need testing.
In this case, such scenarios should be permanently failing on the 2 .4-develop branch without your fix, but they are not. So we don't have any scenario that covers your fix.

**Comment by engcom-Foxtrot** [comment]
Hello, @AntonEvers. Would you like to continue with this PR?

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @engcom-Foxtrot, let's try to use the framework TimeZone and DateTime object instead of generic PHP.

**Comment by engcom-Foxtrot** [comment]
> Hi @engcom-Foxtrot, let's try to use the framework TimeZone and DateTime object instead of generic PHP.

I didn't find any suitable way to do it. Created issue https://github.com/magento/magento2/issues/31968.

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8785](https://jira.corp.magento.com/browse/ENGCOM-8785) has been created to process this Pull Request

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
:x: QA Failed

Switched to the PR and Performed the steps from the Manual Testing scenario
After the last step, still, only one row is displayed in the SQL query
after

**Comment by engcom-Bravo** [comment]
Further investigation discovered that the previous comment https://github.com/magento/magento2/pull/28973#pullrequestreview-594297440 is actually the Expected behavior

:heavy_check_mark: QA Passed

### Manual testing scenario
1. Set time zone to Europe/Amsterdam
2. Set indexers to scheduled
3. Deleta all present Catalog Price Rules
4. Create a Catalog Price Rule for 50% off on all products, but disabled
5. Schedule the price rule to be enabled at a minute or so in the future using catalog staging
6. Wait until catalog stating rules are active and indexer is up to date
7. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`
8. It should output only one row, after all, the rule is applied on all product at the same time.
9. Install [app.zip](https://github.com/magento/magento2/files/4868883/app.zip)
10. Run `bin/magento reindex:product 10`
11. Run this query: `SELECT DISTINCT from_time, to_time FROM catalogrule_product;`

**Actual Result**
Only one row is displayed in the SQL query
after

**Pull request merged**